solr/prometheus-exporter/src/java/org/apache/solr/prometheus/collector/MetricSamples.java

```diff
@@ -19,24 +19,42 @@
 
 import io.prometheus.client.Collector;
 import java.util.HashMap;
+import java.util.HashSet;
 import java.util.List;
 import java.util.Map;
+import java.util.Set;
 import java.util.stream.Collectors;
 
 public class MetricSamples {
 
   private final Map<String, Collector.MetricFamilySamples> samplesByMetricName;
 
+  private final Set<String> sampleMetricsCache;
+
   public MetricSamples(Map<String, Collector.MetricFamilySamples> input) {
     samplesByMetricName = input;
+    sampleMetricsCache = new HashSet<>();
+    for (Collector.MetricFamilySamples metricFamilySamples : input.values()) {
+      addSamplesToCache(metricFamilySamples);
+    }
   }
 
   public MetricSamples() {
     this(new HashMap<>());
   }
 
-  public void addSamplesIfNotPresent(String metricName, Collector.MetricFamilySamples samples) {
-    samplesByMetricName.putIfAbsent(metricName, samples);
+  private void addSamplesToCache(Collector.MetricFamilySamples metricFamilySamples) {
```

Comment:
The diff was confusing to read here.  I think it would be clearer if you move this method to *below* the method that calls it; not before.